src/bech32.h

```diff
@@ -19,11 +19,29 @@
 namespace bech32
 {
 
-/** Encode a Bech32 string. If hrp contains uppercase characters, this will cause an assertion error. */
-std::string Encode(const std::string& hrp, const std::vector<uint8_t>& values);
+enum class Encoding {
+    INVALID, //!< Failed decoding
 
-/** Decode a Bech32 string. Returns (hrp, data). Empty hrp means failure. */
-std::pair<std::string, std::vector<uint8_t>> Decode(const std::string& str);
+    BECH32,  //!< Bech32 encoding as defined in BIP173
```

Comment:
Not necessarily for this PR but we may want to think about renaming this to `BECH32Legacy` or `BECH32v0` or so. In the code right now sometimes "bech32" refers to BIP173 specifically and sometimes to both BIP173 and BIP350. In this line it's the former case, in the namespace it's the latter case. I can see people getting confused by that in the future.

Reply:
I don't that's the right solution. The bech32 module is independent of the concept of segwit or bitcoin or addresses or witness versions. If there are other places where the name bech32 is used in a confusing way to include bech32m, feel free to point them out so they can be fixed (in this PR or elsewhere).

Reply:
Ok, my naming suggestions were just what came to my head at that moment, I did not want to intentionally peg the encoding name to one of its use cases. I just think it's unfortunate that everything including the BIP173 encoding was called "Bech32" and now there is a new encoding called "Bech32m" that is also a Bech32 string. I think it might be helpful to have a naming distinction between Bech32 in general and the BIP173 encoding. But I also hope this code doesn't have to change much in the future and if no one else shares that view, I am happy to have it ignored :)

Reply:
> there is a new encoding called "Bech32m" that is also a Bech32 string.

That's wrong. We should fix places in the code where Bech32m is referred to as Bech32. I know of one, addresstype/OutputType::BECH32, which I'm intentionally ignoring in this to minimize code changes (see earlier discussion), but should be fixed still.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/20861#discussion_r596346647

The name "bech32m" does maybe seem unfortunate. It might have been nice to chose a completely different name ("fatfinger32, now catching more stupid typos!") since from a user perspective you can't use a bech32 address when you need a bech32m address, or vice versa. It's just behind the scenes that they share an extremely similar implementation, and the "human readable" parts can be interpreted the same way.

Reply:
You are hereby promoted to Official Namer Of Stuff.

Reply:
Nice, so much easier than my other jobs!

Reply:
> That's wrong. We should fix places in the code where Bech32m is referred to as Bech32.

Ok, re-reading the module now I think this is mostly clear and I was probably bringing in my own preconceived ideas or looking at older code like the comment 2 lines above that was actually removed. I nit-picked through the module and here a few suggestions where this could be clarified further: https://github.com/fjahr/bitcoin/commit/2156afd69e6e5bd587a9d734b21c3e068a037156 but obviously feel free to ignore or keep for a follow-up together with the OutputType maybe.